src/app/collective-rewards/utils/chartUtils.ts

```diff
@@ -9,3 +19,246 @@ export const formatShort = (n: number) => {
+    }
+
+    // For backing and rewards
+    if (item.day) {
+      const itemTime = convertToTimestamp(item.day)
+      return itemTime >= minDate
+    }
+
+    return false
+  })
+}
+
+/**
+ * Generic function to fill gaps between dates with the last known value
+ * Creates daily data points from sparse data using functional approach
+ */
+const interpolateDataByDay = <T extends { day?: number; currentCycleStart?: string }>(
+  data: T[],
+  getTimestamp: (item: T) => number,
+  createInterpolatedItem: (day: number, lastKnownItem: T) => T,
+): T[] => {
+  const sortedData = [...data].sort((a, b) => getTimestamp(a) - getTimestamp(b))
+  const startDate = getTimestamp(sortedData[0])
+  const endDate = getTimestamp(sortedData[sortedData.length - 1])
+
+  const dataMap = new Map(sortedData.map(item => [getTimestamp(item), item]))
+
+  const totalDays = Math.floor((endDate - startDate) / ONE_DAY_IN_SECONDS) + 1
+  const allDays = Array.from({ length: totalDays }, (_, i) => startDate + i * ONE_DAY_IN_SECONDS)
+
+  let lastKnownItem = sortedData[0]
+
+  return allDays.map(day => {
+    const actualData = dataMap.get(day)
+
+    if (actualData) {
+      lastKnownItem = actualData
+      return actualData
+    }
+
+    return createInterpolatedItem(day, lastKnownItem)
+  })
+}
+
+/**
+ * Fills gaps between dates with the last known value for DailyAllocationItem
+ */
+const interpolateDailyBackingData = (data: DailyAllocationItem[]): DailyAllocationItem[] => {
+  const lastDataPoint = data[data.length - 1]
+  const todayTimestamp = Math.floor(Date.now() / 1000)
+
+  let extendedData = data
+  if (lastDataPoint.day < todayTimestamp) {
+    const syntheticTodayItem: DailyAllocationItem = {
+      ...lastDataPoint,
+      id: `extended-to-today-${todayTimestamp}`,
+      day: todayTimestamp,
+    }
+    extendedData = [...data, syntheticTodayItem]
+  }
+
+  return interpolateDataByDay(
+    extendedData,
+    item => item.day,
+    (day, lastKnownItem) => ({
+      id: `interpolated-${day}`,
+      day,
+      totalAllocation: lastKnownItem.totalAllocation,
+    }),
+  )
+}
+
+/**
+ * Fills gaps between dates with the last known value for CycleRewardsItem
+ */
+const interpolateCycleRewardsData = (data: CycleRewardsItem[], targetEndDate: number): CycleRewardsItem[] => {
+  const lastItem = data[data.length - 1]
+  const extendedTargetEndDate = targetEndDate + ONE_DAY_IN_SECONDS
+
+  const syntheticEndItem: CycleRewardsItem = {
+    ...lastItem,
+    id: `synthetic-end-${extendedTargetEndDate}`,
+    currentCycleStart: extendedTargetEndDate.toString(),
+  }
+
+  const extendedData = [...data, syntheticEndItem]
+
+  return interpolateDataByDay(
+    extendedData,
+    item => Number(item.currentCycleStart),
+    (day, lastKnownItem) => ({
+      ...lastKnownItem,
+      id: `interpolated-${day}`,
+      currentCycleStart: day.toString(),
+    }),
+  )
+}
+
+/**
+ * Calculate USD value from RIF and RBTC rewards
+ */
+const calculateRewardsUSD = (
```

Comment:
Why don't we reuse existing functions for this? Please have a look at `getCombinedFiatAmount` that is used in other places